File: 백준/Silver/15658. 연산자 끼워넣기 （2）/연산자 끼워넣기 （2）.py
import sys

input = sys.stdin.readline
n = int(input())
arr = [int(i) for i in input().split()]
oper = [int(i) for i in input().split()]

calc_fun = [lambda x: x[0] + x[1], lambda x: x[0] - x[1], lambda x: x[0] * x[1], lambda x: int(x[0] / x[1])]


def main():
    li = [-float('inf'), float('inf')]
    def recur(order, calcv):
        if order == len(arr):
            li[0], li[1] = max(li[0], calcv), min(li[1], calcv)
            return
        
        for i in range(4):
            if oper[i] > 0:
                oper[i] -= 1
                recur(order + 1, calc_fun[i]((calcv, arr[order])))
                oper[i] += 1
                
    recur(1, arr[0])
    
    print(li[0])
    print(li[1])

main()

File: 백준/Silver/15658. 연산자 끼워넣기 （2）/연산자 끼워넣기 （2）.js
const fs = require('fs');
const read = fs.readFileSync('dev/stdin').toString().trim()
let [n, arr, oper] = read.split("\n");

arr = arr.split(" ").map(Number);
oper = oper.split(" ").map(Number);
const calcFun = [(a,b) => a+b, (a,b) => a-b, (a,b) => a*b, (a,b) => parseInt(a/b) + 0];

const main = () => {
    const li = [-Infinity, Infinity]
    const recur = (order, calcv) => {
        if(order === arr.length){
            li[0] = Math.max(li[0], calcv)
            li[1] = Math.min(li[1], calcv)
            return
        }
        for(let i = 0; i < 4; i++){
            if (oper[i] > 0){
                oper[i] -= 1
                recur(order + 1, calcFun[i](calcv, arr[order]))
                oper[i] += 1
            }
        }
    }
    
    recur(1, arr[0])
    
    li.forEach((a) => {
        console.log(a);
    })
}

main()